hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/stack/savepoints/BuilderSinkImpl.java

```diff
@@ -0,0 +1,171 @@
+     * Only the {@link com.hedera.node.app.workflows.handle.stack.savepoints.FirstRootSavepoint} uses this
+     * constructor, since other savepoints are limited by just a total capacity that doesn't discriminate
+     * between preceding and following builders.
+     * @param maxPreceding the maximum number of preceding builders
+     * @param maxFollowing the maximum number of following builders
+     */
+    public BuilderSinkImpl(final int maxPreceding, final int maxFollowing) {
+        this.maxPreceding = maxPreceding;
+        this.maxFollowing = maxFollowing;
+        this.maxTotal = maxPreceding == Integer.MAX_VALUE ? Integer.MAX_VALUE : maxPreceding + maxFollowing;
+    }
+
+    /**
+     * Constructs a {@link BuilderSinkImpl} with the given maximum number of total builders that can be constructed
+     * in the savepoint. Used for any savepoint other than the first root savepoint.
+     * @param maxTotal the maximum number of total builders
+     */
+    public BuilderSinkImpl(final int maxTotal) {
+        this.maxPreceding = maxTotal;
+        this.maxFollowing = maxTotal;
+        this.maxTotal = maxTotal;
+    }
+
+    @Override
+    public List<SingleTransactionRecordBuilder> allBuilders() {
+        if (precedingBuilders.isEmpty()) {
+            return followingBuilders;
+        } else {
+            final List<SingleTransactionRecordBuilder> allBuilders = new ArrayList<>(precedingBuilders);
+            allBuilders.addAll(followingBuilders);
+            return allBuilders;
+        }
+    }
+
+    @Override
+    public boolean hasBuilderOtherThan(@NonNull final SingleTransactionRecordBuilder baseBuilder) {
+        requireNonNull(baseBuilder);
+        for (final var builder : precedingBuilders) {
+            if (builder != baseBuilder) {
+                return true;
+            }
+        }
+        for (final var builder : followingBuilders) {
+            if (builder != baseBuilder) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public <T> void forEachOtherBuilder(
+            @NonNull final Consumer<T> consumer,
+            @NonNull final Class<T> builderType,
+            @NonNull final SingleTransactionRecordBuilder baseBuilder) {
+        requireNonNull(builderType);
+        requireNonNull(consumer);
+        requireNonNull(baseBuilder);
+        for (final var builder : followingBuilders) {
+            if (builder != baseBuilder) {
+                consumer.accept(builderType.cast(builder));
+            }
+        }
+        for (final var builder : precedingBuilders) {
+            if (builder != baseBuilder) {
+                consumer.accept(builderType.cast(builder));
+            }
+        }
+    }
+
+    @Override
+    public void flushInOrder(@NonNull final BuilderSink parentSink) {
+        requireNonNull(parentSink);
+        ((BuilderSinkImpl) parentSink).precedingBuilders.addAll(precedingBuilders);
```

Comment:
Code like this often indicates the design could be improved

Reply:
Yes, added method to `BuilderSink` and removed cast